### ISSUE TITLE
SendKeys does not 'send keys' but rather sets or appends text

### ISSUE THREAD

**Issue by DGCBio** [appium/appium-uiautomator2-server#136]
In looking at the source code, this seems like it could be a limitation of the underlying Android classes, but it is certainly not 'sending keys' in the sense of entering them one at a time.  This has a real effect on our testing as one thing we are trying to test is whether or not a restriction on number of characters a text box allows is being enforced.  

Has there been any consideration for this?  A proper sendkeys would be welcome for our team.

**Comment by jlipps** [comment]
Yes, unfortunately the underlying classes only allow setting or clearing of text, not entering one character at a time. It's not webdriver spec-compliant for this reason.

Presumably however even doing things this way would respect field length limits, so you could still make that assertion on your side? (If necessary, trying with successively longer substrings)

**Comment by DGCBio** [comment]
We will probably just revert these methods to use adb.  

Once the SendKeys tried to append characters to the current string past the restriction, the entire text was not entered (because of the currentText + newText code in this repository), which meant that, yes our restriction was working, but it was unexpected behavior and we didn't want it to be in the API for our automated test writers to use.

Thanks for the prompt reply!

**Comment by jlipps** [comment]
Makes sense. We'll keep looking for a way to do things character-by-character

**Comment by olezhkalviv** [comment]
@jlipps 
We have similar problem. We have an application that allows user to login via form and using microsoft open id. 

So for open id the application opens browser where on microsoft login site you need to: enter email - tap continue - enter password - tap continue - and you got back to app where you are authentificated.

With previous version it worked fine, but with uiautomator2 when you SetText or SetImmediateValue (c#) for email and tap continue it asks to enter email (same for password). 

Email is entered in a field correctly. To go to next step you need manually to add 1 space and delete it, or add symbol and delete, or delete last char and enter it (manually from keyboard). There is no way to make somewthing similar from script.

Is there any possibility how to handle that? Looks like when uiautomator2 setting text for a field it doesnt update page that something is inserted and page still thinks that field is empty.
![ddddddd](https://user-images.githubusercontent.com/25117767/40074451-54fa1fea-5882-11e8-9e61-17e6eb36cb62.png)

**Comment by olezhkalviv** [comment]
@mykola-mokhnach 

I was able to find workaround previously: 
```
Android.PasswordOpenId.Tap();
Process.Start("cmd", $"/C adb shell input text {password}");
```
and it was working
with your suggestion I made sth like:
```
Android.PasswordOpenId.SendKeys(password);
Manager.Driver.PressAndroidKeyCode(AndroidKeyCode.Space);
Manager.Driver.PressAndroidKeyCode(AndroidKeyCode.BackSpace);
```
thanks, it is better as it doesnt require cmd launch

The same for tapping - for some reason it doesn't work as element.tap or driver.tap(element) or driver.tap(x, y). Used workarodund:
`Process.Start("cmd", $"/C adb shell input tap {Android.NextOpenId.Location.X} {Android.NextOpenId.Location.Y}");`

but now:
```
Android.NextOpenId.Tap();
Manager.Driver.PressAndroidKeyCode(AndroidKeyCode.Enter);
```
